src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,99 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+import org.apache.commons.lang.ObjectUtils;
+
 public class ArrayList<T> implements List<T> {
+    private static int capacity = 10;
+    private static final double RESIZE_FACTOR = 1.5;
+    private int size;
+    private T[] elements = (T[]) new Object[capacity];
+
     @Override
     public void add(T value) {
-
+        checkSize();
+        elements[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("out of bound exception");
+        }
+        checkSize();
+        System.arraycopy(elements, index, elements, index + 1, size - index);
+        elements[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("out of bound exception");
+        }
+        return elements[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("out of bound exception");
+        }
+        elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("out of bound exception");
+        }
+        int updateSize = size - 1;
+        final T remove = elements[index];
+        System.arraycopy(elements, index + 1, elements, index, updateSize - index);
+        elements[size - 1] = null;
+        size--;
+        return remove;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        remove(getIndexByElement(element));
+        return element;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkSize() {
```

Comment:
It is better to come up with a more informative and suitable name

Reply:
Agree. `checkSizeAndGrowIfNeeded` or `growIfNeeded` sounds better